### PR TITLE
Add support for basic UDP networking

### PR DESCRIPTION
**Pull request by Noah-Kennedy**
This PR adds support for a few basic UDP operations. More will be added later, but I'd like to get some discussion going about the implementation details used here before more work is done.

### CODE REVIEW

File: src/net/udp.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+use crate::buf::{IoBuf, IoBufMut};
+use crate::driver::{Op, SharedFd};
+
+/// An `io_uring`-backed implementation of the User Datagram Protocol
+pub struct UdpSocket {
+    fd: SharedFd,
+    _inner_sock: ManuallyDrop<std::net::UdpSocket>,
+}
+
+impl UdpSocket {
```
Comment (carllerche) [inline src/net/udp.rs:16]: There should also be an explicit `close` fn, similar to the `File` type.

**Review by carllerche — COMMENTED** [review]
Looks good, I only did a quick skim so far. I left some notes inline. Ideally there would be more tests, you can inspire yourself from the File tests.

I also didn't run the tests locally yet, as mentioned on Discord, it would be nice to get CI setup now that Github action supports it.

File: src/net/udp.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+use tokio::net::ToSocketAddrs;
+
+use crate::buf::{IoBuf, IoBufMut};
+use crate::driver::{Op, SharedFd};
+
+/// An `io_uring`-backed implementation of the User Datagram Protocol
+pub struct UdpSocket {
+    fd: SharedFd,
+    _inner_sock: ManuallyDrop<std::net::UdpSocket>,
```
[the remark this answers is not in the document]
Comment (Noah-Kennedy) [reply src/net/udp.rs:13]: I am only keeping this around to add methods like `peer_addr` and `local_addr` in the future.

File: src/net/udp.rs (hunk 1 above)
Comment (Noah-Kennedy) [reply src/net/udp.rs:16]: There are a lot of functions this needs added, I was planning to add a bunch like close, send_to, etc, in another PR.

File: src/net/udp.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+impl UdpSocket {
+    /// Bind a socket to specified address.
+    pub async fn bind<A>(addr: A) -> IoResult<Self>
+    where
+        A: ToSocketAddrs,
```
[the remark this answers is not in the document]
Comment (Noah-Kennedy) [reply src/net/udp.rs:20]: I'm inclined to keep this the case for now for consistency, but I see your point here.

**Comment by Noah-Kennedy** [comment]
I'm going to close this once LinkTed's PR is in.

**Comment by Noah-Kennedy** [comment]
Obseleted by #40

**Pull request closed without merging**